src/reducers/secrets.js

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { combineReducers } from 'redux';
+import keyBy from 'lodash.keyby';
+
+function byId(state = {}, action) {
+  switch (action.type) {
+    case 'SECRETS_FETCH_SUCCESS':
+      return keyBy(action.data, 'name');
```

Comment:
should this be `id`?

Reply:
currently when secrets are requested from the API, the object returned doesn't contain an ID field so therefore I use the name to organize them.

Reply:
ok so the reducer should be `byName` then? and the delete action should specify a `name` rather than an `id` for consistency

Reply:
I'm still a little confused on this one, the `byNamespace` reducer is using a `uid` but looking at the `fetchSecrets` action creator this just seems to be the index of the secret in the array returned by the API. There's no guarantee this will be consistent across requests and so it should not be used.

The intent of the `byNamespace` and `byId` reducers is to allow for consumers to easily get the required resources for a number of use cases:
1. names of resources of a given type for a specific namespace
2. values of resources of a given type for a specific namespace
3. all resources of a given type
4. a single resource of a given type by id

while normalizing the data to avoid duplication. This is achieved by storing the `name` and `uid` of the resource through the `byNamespace` reducer, and the full resource object through the `byId` reducer.

Given a resource `{ name: 'myResource', namespace: 'default', uid: '1234', ... }` it would be stored as:

```
byId: {
   1234: { name: 'myResource', namespace: 'default', uid: '1234', ... }
}
byNamespace: {
  default: {
    myResource: '1234'
  }
}
```

This allows us to provide the data required for the cases above:
1. `Object.keys(byNamespace[namespace])`
2. `Object.values(byNamespace[namespace])` gives the `uid`s, which are then used to retrieve the corresponding values from `byId`
3. `Object.values(byId)`
4. `byId[id]`


Secrets do not have a uid, and their names are not guaranteed to be unique across namespaces. You may be able to achieve your needs with just a `byNamespace` reducer:

```
byNamespace: {
  [name]: secret
}
```

I'd start by documenting how they're expected to be used in the front end, identifying the different use cases, then model the data in a way that makes sense for those uses.